adr/2025-11-03-improved-http-cache-layer.md

```diff
@@ -0,0 +1,62 @@
+* **Actual cache-control configuration is hard-coded and splattered**: The values set for `cache-control` headers are hard-coded and splattered (e.g., hardly configured in the reverse proxy config and in shopware code), they cannot be configured based on projects needs, and also on route level only the max-age is configurable.
+
+## Decision
+
+We will rework the HTTP-cache layer to address the limitations mentioned above, for that the following changes will be made:
+
+### Only use cache-relevant rule-ids inside cache-hash
+
+Most rule-ids are not relevant for caching, e.g., when they are not used at all, or only for checkout-related features (e.g., payment-methods, promotions).
+Therefore, we will only include the rule-ids that are relevant for caching in the cache-hash, we do the distinction based on the existing `rule-areas`.
+By default, all rule-ids, that are used in the `product` area, will be included in the cache-hash. For extensibility, a new event will be introduced that allows modifying the list of rule-areas that are relevant for caching.
+
+### Configurable caching policies
+
+The caching policies used will be moved from the code directly to the configuration. We will allow defining default caching policies based on the area (storefront or store-api), as well as use-case (cached or uncached).
+We also allow route level configuration that will override the default caching policies.
+
+### Simplify reverse proxy configuration
+
+We will simplify the reverse proxy configuration by only relying on the `sw-cache-hash`, as the only application state that the reverse proxy needs to take into account.
+The `sw-cache-hash` will be set as a cookie and as a header, additionally we add it as a `vary` header to the response.
+Adding it as header allows the default `vary` header implementation of reverse proxies to work.
+The cookie is used to make it easier for clients to pass along the correct hash with all further requests, without the need to manually handle the header.
+The only shopware specific reverse proxy configuration will be to set the `sw-cache-hash` header based on the `sw-cache-hash` cookie on the reverse proxy.
+
+Additionally, the use of clear and configurable policies for the cache-control headers will remove the need to manually override the cache-control headers on the reverse proxy side.
+
+### HTTP-Cache support for store-api
+
+We will add HTTP-Cache support for the store-api as well. The caching behaviour and used patterns are the same as for the storefront.
+So the configuration (on shopware side, as well on the reverse proxy) will be the same as in the storefront.
+To make the caching applicable for the store-api we will adjust the store-api routes to support `GET` requests where it makes sense, clients should preferably use the `GET` requests.
+
+## Consequences
+
+### Storefront pages for logged-in users and filled carts are cached by default
```

Comment:
Those are the three distinct breaks we are doing, i was thinking if we should have a dedicated feature flag for each of the breaks or if we want to have it behind a single caching feature flag 🤔